keras/applications/imagenet_utils.py

```diff
@@ -35,26 +36,21 @@ def preprocess_input(x, data_format=None, mode='caffe'):
         x *= 2.
         return x
 
+    vgg_mean = [103.939, 116.779, 123.68]
     if data_format == 'channels_first':
         if x.ndim == 3:
             # 'RGB'->'BGR'
             x = x[::-1, ...]
             # Zero-center by mean pixel
-            x[0, :, :] -= 103.939
-            x[1, :, :] -= 116.779
-            x[2, :, :] -= 123.68
+            x = x - np.array(vgg_mean, dtype=np.float32).reshape((3, 1, 1))
```

Comment:
The problem with this approach is that you are creating (and allocating) a new Numpy array for every function call (and it gets called for every new image your network sees). Better to make it a global variable (lazily instantiated upon first use) I suppose?

Reply:
I guess I understand how to make it a global variable but am not sure how to lazily instantiate it. 
I looked around in the code base hoping to find an example but could not. Can you please provide some guidance?

By the way, this numpy arrays are rather small:
```sh
>>> vgg_mean = [103.939, 116.779, 123.68]
>>> np.array(vgg_mean, dtype=np.float32).reshape((3, 1, 1))
array([[[ 103.93900299]],

       [[ 116.77899933]],

       [[ 123.68000031]]], dtype=float32)
```
therefore, can we afford to instantiate global np arrays when the module is imported?

Reply:
Basically, you'd do:

```python
_IMAGENET_MEAN_CL = None
_IMAGENET_MEAN_CF = None

def preprocess():
    global _IMAGENET_MEAN_CL
    global _IMAGENET_MEAN_CF
    if channels_first:
       if _IMAGENET_MEAN_CF is None:
          _IMAGENET_MEAN_CF = np.array(...)
```

Reply:
Whether or not we can afford it, we should never run code upon import (like calling `np` functions). May seem innocuous, but in practice there are always unforeseen side effects.